### PR TITLE
[Automation] Adding vagrant files

### PR DESCRIPTION
**Pull request by Yarboa**
Adding vagrant environment for this repo
Moving vagrant scripting into vagrant directory
Pipfiles in vagrant directory
Adding vagrant README

Updated check.yaml not to fail in github actions

### CODE REVIEW

**Review by EdDev — COMMENTED** [review]
Can we move all the files from the project root to a separated folder (e.g. `vagrant`)?

I prefer the base to be targeting containers (or self managed env) and as a companion for "backward compatibility" to have vagrant as an option.

WDYT?

**Comment by EdDev** [comment]
Also, can you please change the commit message and the subject of the PR to something like: `env: Support deployment using Vagrant`

**Comment by Yarboa** [comment]
> Also, can you please change the commit message and the subject of the PR to something like: `env: Support deployment using Vagrant`

:+1:

**Comment by Yarboa** [comment]
> Can we move all the files from the project root to a separated folder (e.g. `vagrant`)?
> 
No need, for now

> I prefer the base to be targeting containers (or self managed env) and as a companion for "backward compatibility" to have vagrant as an option.
>

I will give it extra thought

> WDYT?

**Comment by Yarboa** [comment]
Need to update README for vagrant

File: README.vagrant.md [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+### Tests
+[pytest](https://docs.pytest.org) is used to run tests.
+
+### Arbitrary commands
+During development, there may be a need to run various command,
+like creating applications and running migrations.
+
+These steps can be done by either accessing the container terminal,
+or adding the command at the tail (see below).
+
```
Comment (Yarboa) [inline README.vagrant.md:61]: Please delete, remains from previous readme?

**Review by EdDev — COMMENTED** [review]
Before I proceed in reading everything, I wanted to give a initial high level feedback.
I am fine with taking it without these points and then changing it myself, but if you have the time, it will be nicer.

- Currently the project does not use `pipenv`, for simplicity. But it is probably wrong and it makes sense to use it. So if you can add it in a separate PR/commit, it will be nice. We will not need this in this `vagrant` folder.
- The github action workflow will just require modification to align with the previous point. We will not need another one for sure.

Thank you very much for the investment and help in this!

File: README.md [hunk 2]
```diff
@@ -13,5 +15,13 @@ Dedicated instruction on how to build and use the project exist for:
 [1] The instructions should also work with `docker`, but it has not
 been formally tested.
 
+### Virtualization
+The project could be also targeted with Vagrant
```
Comment (EdDev) [inline README.md]: Missing `.` at the end of the sentence.

File: README.md [hunk 3]
```diff
@@ -13,5 +15,13 @@ Dedicated instruction on how to build and use the project exist for:
...
+
+Dedicates instructions of how to set vagrant environement exist for:
+- [vagrant](README.vagrant.md) [2]
```
Comment (EdDev) [inline README.md]: `[2]` is suppose to point to a note?

File: README.vagrant.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+# Overview
+
+The instructions describe in this document focus on the usage of vagrant,
+a tool that build and manage vms.
```
Comment (EdDev) [inline README.vagrant.md]: vms ==> `VM/s`

**Comment by Yarboa** [comment]
> Before I proceed in reading everything, I wanted to give a initial high level feedback. I am fine with taking it without these points and then changing it myself, but if you have the time, it will be nicer.
> 
> * Currently the project does not use `pipenv`, for simplicity. But it is probably wrong and it makes sense to use it. So if you can add it in a separate PR/commit, it will be nice. We will not need this in this `vagrant` folder.

You are correct,
But  IIUC the issue is look and feel from the students perspective,
We could add some comment related to pipenv, WDYT?

> * The github action workflow will just require modification to align with the previous point. We will not need another one for sure.
> 
> Thank you very much for the investment and help in this!

File: README.md (hunk 3 above)
Comment (Yarboa) [reply README.md]: Ack

File: README.vagrant.md (hunk 4 above)
Comment (Yarboa) [reply README.vagrant.md]: Ack

**Comment by Yarboa** [comment]
ToDo,
Use export for Pipfile 
export PIPENV_PIPFILE=./vagrant/Pipfile
Instead of links. and it should be helpful with CI, also

File: README.md (hunk 2 above)
Comment (Yarboa) [reply README.md]: Ack

File: .github/workflows/check.yaml [hunk 5]
```diff
@@ -16,11 +16,16 @@ jobs:
...
+        uses: actions/checkout@v2
+      - name: Setup Python
+        uses: actions/setup-python@v3
         with:
-          options: "--check --diff -S"
-          version: "22.1.0"
+          python-version: ${{ env.PYTHON_VERSION }}
+      - name: Install dependencies
+        run: pip install -r requirements.txt
+      - name: Run flake8
```
Comment (EdDev) [inline .github/workflows/check.yaml]: I am unclear why the format job got changed.
It does not seem to be related to vagrant.

If we want to delve into details, it is not clear why we need to install all dependencies.
Also, the step is running `black` and not `flake8`.

File: README.md [hunk 6]
```diff
@@ -13,5 +15,13 @@ Dedicated instruction on how to build and use the project exist for:
 [1] The instructions should also work with `docker`, but it has not
 been formally tested.
 
+### Virtualization
+The project could be also targeted with Vagrant.
```
Comment (EdDev) [inline README.md]: > also targeted with

be deployed with

File: README.vagrant.md [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+cd django-testing-tutorial
+```
+
+To build the virtualization image, need to run vagrant up  
+`cd vagrant`  
+`vagrant up`  
+
+*Note:* all vagrant commands at host level should run from project/vagrant directory
+
+Wait vagrant output as following:
```
Comment (EdDev) [inline README.vagrant.md]: Wait for the vagrant output to appear:

File: vagrant/app_setup.sh [hunk 8]
```diff
@@ -8,9 +8,12 @@
 sudo -n dnf install -y pipenv
 
 cd /vagrant/vagrant
+export PIPENV_PIPFILE=/vagrant/vagrant/Pipfile
+mv /vagrant/requirements{.txt,}
```
Comment (EdDev) [inline vagrant/app_setup.sh:12]: Could you please explain in the commit message or here in a comment, why it is needed?

**Review by EdDev — CHANGES_REQUESTED** [review]
Thank you very much!
I had a few comments.

2nd commit: `[Updae to cloudbase fedora-36]`
Has a typo.

File: README.md (hunk 6 above)
Comment (Yarboa) [reply README.md]: Ack

File: README.vagrant.md (hunk 7 above)
Comment (Yarboa) [reply README.vagrant.md]: Ack

File: vagrant/app_setup.sh (hunk 8 above)
Comment (Yarboa) [reply vagrant/app_setup.sh:12]: some, how even if am setting PIPENV PIPENV_PIPFILE=/vagrant/vagrant/Pipfile

once there is a call, using pipenv run python manage.py, 
pipenv find the requirements.txt under the directory and installing all the content of the file.

Maybe I should propose a MR to pipenv.
As per now, the container run with centos8 and python 3.8
on vagrant fedora-36 with python3.8


Please advise here, @EdDev 
Two oprions, 
* prepare the container with fedora-36 use the same python.
* research pipenv to understand automatic relation with requirements.txt and maybe propose a PR.

Third options somewhere in the middle
 
 
 I will try this solution also
 https://pipenv-fork.readthedocs.io/en/latest/basics.html#pipfile-lock-security-features

File: .github/workflows/check.yaml (hunk 5 above)
Comment (Yarboa) [reply .github/workflows/check.yaml]: thanks for that, copy paste issues, sorry

**Review by Yarboa — COMMENTED** [review]
@EdDev

File: vagrant/app_setup.sh (hunk 8 above)
Comment (EdDev) [reply vagrant/app_setup.sh:12]: Thanks for the clarification.
I think we can work on this as a follow up, no need to block on it.

**Review by EdDev — DISMISSED** [review]
Thank you!

If you do not mind, It would be nice to squash all the commits. Some are fixing prev commits issues.

**Review by EdDev — APPROVED** [review]
Great! Thank you.

**Pull request merged**